modules/tftp/server.rb

```diff
@@ -90,21 +107,86 @@ def pxe_default
+        return
+      end
+
+      bootloader_path = File.join(bootloader_universe, 'pxegrub2', os, version, arch)
+
+      logger.debug "TFTP: Checking bootloader universe for suitable bootloader directory for"
+      logger.debug "TFTP:   * Operating system: #{os}"
+      logger.debug "TFTP:   * Version: #{version}"
+      logger.debug "TFTP:   * Architecture: #{arch}"
+      logger.debug "TFTP: Checking bootloader universe if \"#{bootloader_path}\" exists."
+      unless Dir.exist?(bootloader_path)
+        logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+
+        bootloader_path = "#{bootloader_universe}/pxegrub2/#{os}/default/#{arch}"
+        logger.debug "TFTP: Checking if fallback directory at \"#{bootloader_path}\" exists."
+        unless Dir.exist?(bootloader_path)
+          logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+          return
+        end
+      end
+
+      bootloader_path
+    end
+
+    def setup_bootloader(mac, os, major, minor, arch, bootfilename_efi, build)
+      pxeconfig_dir_mac = pxeconfig_dir(mac)
+
+      logger.debug "TFTP:   => Deploying host specific bootloader files to \"#{pxeconfig_dir_mac}\""
+
+      FileUtils.mkdir_p(pxeconfig_dir_mac)
+      FileUtils.rm_f(Dir.glob("#{pxeconfig_dir_mac}/*.efi"))
+
+      version = (minor && !minor.empty?) ? "#{major}.#{minor}" : major
+      bootloader_path = bootloader_path(os, version, arch)
+
+      if bootloader_path
+        logger.debug "TFTP: Copying bootloader files from bootloader universe:"
+        logger.debug "TFTP:   - \"#{bootloader_path}/*\" => \"#{pxeconfig_dir_mac}/\""
+        FileUtils.cp_r("#{bootloader_path}/.", "#{pxeconfig_dir_mac}/", remove_destination: true)
+      else
+        pxeconfig_dir = pxeconfig_dir()
+
+        logger.debug "TFTP: Copying default bootloader files:"
+        logger.debug "TFTP:   - \"#{pxeconfig_dir}/grub#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/grub#{bootfilename_efi}.efi\""
+        logger.debug "TFTP:   - \"#{pxeconfig_dir}/shim#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/shim#{bootfilename_efi}.efi\""
+        logger.debug "TFTP:   - \"#{pxeconfig_dir}/grub#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/boot.efi\""
+        logger.debug "TFTP:   - \"#{pxeconfig_dir}/shim#{bootfilename_efi}.efi\" => \"#{pxeconfig_dir_mac}/boot-sb.efi\""
+        FileUtils.cp_r("#{pxeconfig_dir}/grub#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/grub#{bootfilename_efi}.efi", remove_destination: true)
+        FileUtils.cp_r("#{pxeconfig_dir}/shim#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/shim#{bootfilename_efi}.efi", remove_destination: true)
+        FileUtils.cp_r("#{pxeconfig_dir}/grub#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/boot.efi", remove_destination: true)
+        FileUtils.cp_r("#{pxeconfig_dir}/shim#{bootfilename_efi}.efi", "#{pxeconfig_dir_mac}/boot-sb.efi", remove_destination: true)
```

Comment:
I always like a debug line that maps exactly to one operation, so you know what failed. That means `logger.debug` before `FileUtils.cp_r`. So how about:

```ruby
dir = pxeconfig_dir
to_copy = { 
  File.join(dir, "grub#{bootfilename_efi}.efi") => File.join(pxeconfig_dir_mac, "grub#{bootfilename_efi}.efi"),
  File.join(dir, "shim#{bootfilename_efi}.efi") => File.join(pxeconfig_dir_mac, "shim#{bootfilename_efi}.efi"),
  File.join(dir, "grub#{bootfilename_efi}.efi") => File.join(pxeconfig_dir_mac, "boot.efi"),
  File.join(dir, "shim#{bootfilename_efi}.efi") => File.join(pxeconfig_dir_mac, "boot-sb.efi"),
}

to_copy.each do |source, destination|
  logger.debug "TFTP: Copying default bootloader file '#{source}' => '#{destination}'"
  FileUtils.cp_r(source, destination, remove_destination: true)
end
```
```